### ISSUE TITLE
test_target_data_use_device_ptr not standard compliant?

### ISSUE THREAD

**Issue by ams-cs** [SOLLVE/sollve_vv#15]
The test_target_data_use_device_ptr test-case places "use_device_ptr" before "map". This ordering fails with GCC 9, because the implementation requires the "map" first.

According to OpenMP 4.5: "A list item in a use_device_ptr clause must have a corresponding list item in the device data environment." (Section 2.10.1)

This does not say explicitly that the ordering within a directive matters, but does imply that the "use_device_ptr" should not occur until the variable has been mapped. This was the interpretation taken by the GCC implementation.

OpenMP 5.0 appears to agree with this interpretation because new language has been added to cover the case: "If one or more of the use_device_ptr or use_device_addr clauses and one or more map clauses are present on the same construct, the address conversions of use_device_addr and use_device_ptr clauses will occur as if performed after all variables are mapped according to those map clauses."

GCC 10 accepts the test-case, as is, but that's because the implementation has been updated to OpenMP 5.0 (in this regard).

I believe this test-case needs to have those clauses reversed to be truly considered 4.5 compliant.

**Comment by josemonsalve2** [comment]
Hi Andrew, 

I think that even though 4.5 does not say that explicitly, the description of directives in OpenMP says the order of constructs should not matter. 

In 4.5, section 2.1 Directive format. Page 27 line 10
"The order in which clauses appear on directives is not significant."

It is a good point though. But I am not sure if the wording you are describing means that the order in which the clauses appear matters.

**Comment by tob2** [comment]
_"I think that even though 4.5 does not say that explicitly, the description of directives in OpenMP says the order of constructs should not matter."_

One can consider this part of the problem: independent what the user wrote, the compiler can always map first, always use_device_ptr first, process in order, process in reverse order or does so depending of the phase of the moon.

What 4.5 explicitly states: the item must be be **in** "device data environment" – and the only way to do this reliably is by doing so in a **separate** directive.

Additionally, if it were obvious enough that it should work as is, OpenMP 5 had not a very lengthy clause added which explicitly spells out that the mapping happens first. – _I assume the issue only became apparent after OpenMP 4.5 was used; hence, no one thought about spelling it out in the spec before 4.5 was released._

Hence, something like the following should work – the dummy is needed as OpenMP 4.5 requires at least one mapping per "omp target data". (Also in this regard, OpenMP 5 is better.)

```
--- test_target_data_use_device_ptr.c.orig      2019-10-17 09:25:48.856929341 +0200
+++ test_target_data_use_device_ptr.c   2019-10-17 09:28:39.549070825 +0200
@@ -6,3 +6,3 @@
 int main() {
-  int errors = 0, len = 10000, isHost = 0;
+  int errors = 0, len = 10000, isHost = 0, dummy;
   int *array_device = NULL;
@@ -16,5 +16,8 @@ int main() {
 
-#pragma omp target data map(tofrom: array_device[0:len])   \
-    use_device_ptr(array_device)
+#pragma omp target data map(tofrom: array_device[0:len])
   {
+    // Use separate directive for use_device_ptr as only OpenMP 5 guarantees
+    // that the mapping happens before the use_device_{ptr,addr}
+#pragma omp target data map(dummy) use_device_ptr(array_device)
+   {
 #pragma omp target is_device_ptr(array_device) map(tofrom: array_host[0:len]) map(tofrom: isHost)
@@ -27,2 +30,3 @@ int main() {
     } // end target
+   } // end target data
   } // end target data
```

**Comment by spophale** [comment]
Thanks for suggesting the work-around.
I agree, the only way to honor both i)the item must be be in "device data environment" – and the only way to do this reliably is by doing so in a separate directive AND ii)The order in which clauses appear on directives is not significant is to have the mapping on different directive.

**Comment by josemonsalve2** [comment]
This issue seems to be solved by PR #28 Closing it.
